sklearn/tree/_utils.pyx

```diff
@@ -301,3 +303,328 @@ cdef class PriorityHeap:
         self.heap_ptr = heap_ptr - 1
 
         return 0
+
+# =============================================================================
+# WeightedPQueue data structure
+# =============================================================================
+
+cdef class WeightedPQueue:
+    """A priority queue class, always sorted in increasing order.
+
+    Attributes
+    ----------
+    capacity : SIZE_t
+        The capacity of the array
+
+    array_ptr : SIZE_t
+        The water mark of the array; the array grows from left to right in the
+        array ``array_``. array_ptr is always less than capacity.
+
+    array_ : WeightedPQueueRecord*
+        The array of array records. The minimum element is on the left;
+        the array grows from left to right
+    """
+
+    def __cinit__(self, SIZE_t capacity):
+        self.capacity = capacity
+        self.array_ptr = 0
+        self.array_ = <WeightedPQueueRecord*> calloc(capacity, sizeof(WeightedPQueueRecord))
+
+        if self.array_ == NULL:
+            raise MemoryError()
+
+    def __dealloc__(self):
+        free(self.array_)
+
+    cdef bint is_empty(self) nogil:
+        return self.array_ptr <= 0
+
+    cdef SIZE_t size(self) nogil:
+        return self.array_ptr
+
+    cdef int push(self, DOUBLE_t data, DOUBLE_t weight) nogil:
+        """Push record on the array.
+        Returns 0 if successful; -1 on out of memory error.
+        """
+        cdef SIZE_t array_ptr = self.array_ptr
```

Comment:
But if you're moving all values over anyway, there might not be an actual speed gain from finding the position faster since you still have to bubble things? Maybe run a time test